opentasks-provider/src/main/java/org/dmfs/provider/tasks/processors/instances/Validating.java

```diff
@@ -43,50 +69,79 @@ public Validating(EntityProcessor<InstanceAdapter> delegate)
     public InstanceAdapter insert(SQLiteDatabase db, InstanceAdapter entityAdapter, boolean isSyncAdapter)
     {
         validate(entityAdapter, isSyncAdapter);
-        return mDelegate.insert(db, entityAdapter, isSyncAdapter);
+
+        Optional<Long> instanceId = new NullSafe<>(entityAdapter.taskAdapter().valueOf(TaskAdapter.ORIGINAL_INSTANCE_ID));
+        if (instanceId.isPresent())
+        {
+            String timeStampString = Long.toString(entityAdapter.taskAdapter().valueOf(TaskAdapter.ORIGINAL_INSTANCE_TIME).getTimestamp());
+            // there better be no instance for this yet, otherwise this is going to fail
+            try (Cursor c = db.query(
+                    TaskDatabaseHelper.Tables.INSTANCE_VIEW,
+                    new String[] { TaskContract.Instances._ID },
+                    // find any instance which refers to the given original ID and has the same instance time
+                    // for recurring tasks this matches the INSTANCE_ORIGINAL_TIME, for non-recurring tasks this matches start or due (whichever is present).
+                    String.format("(%1$s == ? or %2$s == ?) and (%3$s == ? or %3$s is null and %4$s == ? or %3$s is null and %4$s is null and %5$s == ?) ",
+                            TaskContract.Instances.TASK_ID,
+                            TaskContract.Instances.ORIGINAL_INSTANCE_ID,
+                            TaskContract.Instances.INSTANCE_ORIGINAL_TIME,
+                            TaskContract.Instances.INSTANCE_START,
+                            TaskContract.Instances.INSTANCE_DUE),
+                    new String[] {
+                            instanceId.value().toString(),
+                            instanceId.value().toString(),
+                            timeStampString,
+                            timeStampString,
+                            timeStampString },
+                    null,
+                    null,
+                    null))
+            {
+                if (c.getCount() > 0)
+                {
+                    throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "Instance %s of task %d already exists",
```

Comment:
Wouldn't `IllegalArgumentException` or `IllegalStateException` be more appropriate?

Reply:
I've been thinking the same, but I wasn't sure. I'll change it.